// onebrick-node-sdk/src/core/worker/transactionWorker.ts
import axios from "axios";
import { ENDPOINT } from "../util/constant";
import { Configuration } from "../util/configuration";
export class TransactionWorker {
  accessToken: String = "";
  configuration: Configuration;
  constructor(configuration:Configuration,accessToken: string) {
    this.accessToken = accessToken;
    this.configuration = configuration;
  }

  requestTransactionGeneral(from: string, to: string): Promise<any> {
    return new Promise<any>(async (resolve, reject) => {
      const url = await this.configuration?.getAPIUrl();
      const headers = {
        "Content-Type": "application/json",
        "X-Channel-Id": "SDK NodeJS V 1.01",
        Authorization: `Bearer ${this.accessToken}`,
      };
      axios
        .get(`${url}/${ENDPOINT.TRANSACTION_DATA_GENERAL}?from=${from}&to=${to}`, { headers: headers })
        .then((data: any) => {
          resolve(data.data);
        })
        .catch((err: Error) => {
          reject(err);
        });
    });
  }
  requestTransactionGeneralMFA(from: string, to: string): Promise<any> {
    return new Promise<any>(async (resolve, reject) => {
      const url = await this.configuration?.getAPIUrl();
      const headers = {
        "Content-Type": "application/json",
        "X-Channel-Id": "SDK NodeJS V 1.01",
        Authorization: `Bearer ${this.accessToken}`,
      };
      axios
        .post(`${url}/${ENDPOINT.TRANSACTION_DATA_GENERAL_MFA}`, { headers: headers })
        .then((data: any) => {
          resolve(data.data);
        })
        .catch((err: Error) => {
          reject(err);
        });
    });
  }
}
