// onebrick-node-sdk/src/core/models/authentication.ts
export interface Authentication {
  status: number;
  message: string;
  data: AuthenticationData;
}

export interface AuthenticationData {
  clientId: number;
  clientName: string;
  clientFullName: string;
  clientEmail: string;
  clientAlias: string;
  clientImageUrl: string;
  clientFavicon: string;
  redirectRefId: number;
}

export interface AuthenticationRequest {
  status: number;
  message: string;
  data: AuthenticationRequestData;
}

export interface AuthenticationRequestData {
  access_token: string;
}

export interface AuthenticationRequestPayload {
  accessToken: string;
  userId: string;
  redirectUrl: string;
}

export interface AuthenticationRequestResponse {
  status: number;
  message: string;
  data: AuthenticationRequestResponseData;
}

export interface AuthenticationRequestResponseData {
  clientId: number;
  clientName: string;
  clientFullName: string;
  clientEmail: string;
  clientAlias: string;
  clientImageUrl: string;
  clientFavicon: string;
  redirectRefId: number;
}

export interface AuthenticationRequestBankTypePayload {
  institutionId?: number;
  institution_id?: number;
  sessionId?: string;
  username?: string;
  password?: string;
  userId?: string;
  token?: string;
  corporateId?: string;
  redirectRefId?: number;
}
export interface AuthenticationRequestEcommerceTypePayload {
  institutionId: number;
  username: string;
  password: string;
  redirectRefId?: number;
  sessionId?: string;
  requestId: string;
  token?: string;
}

export interface AuthenticationRequestEwalletType2Payload {
  username?: string;
  refId?: string;
  otpNumber?: string;
  pin?: number;
  deviceId?: string;
  redirectRefId?: number;
}
export interface AuthenticationRequestEwalletTypePayload {
  institutionId: number;
  username: string;
  otp?: string;
  otpToken?: string;
  sessionId?: string;
  uniqueId?: string;
  redirectRefId: number;
}
