// onebrick-node-sdk/dist/core/worker/ecommerceDataWorker.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcommerceDataWorker = void 0;
const axios_1 = __importDefault(require("axios"));
const constant_1 = require("../util/constant");
class EcommerceDataWorker {
    constructor(configuration, accessToken) {
        this.accessToken = "";
        this.accessToken = accessToken;
        this.configuration = configuration;
    }
    requestEcommerceBuyerIdentity() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const url = yield ((_a = this.configuration) === null || _a === void 0 ? void 0 : _a.getAPIUrl());
            const headers = {
                "Content-Type": "application/json",
                "X-Channel-Id": "SDK NodeJS V 1.01",
                Authorization: `Bearer ${this.accessToken}`,
            };
            axios_1.default
                .get(`${url}/${constant_1.ENDPOINT.ECOMMERCE_BUYER_IDENTITY}`, { headers: headers })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
    requestEcommerceBuyerTransaction(numberOfTransaction) {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const url = yield ((_a = this.configuration) === null || _a === void 0 ? void 0 : _a.getAPIUrl());
            const headers = {
                "Content-Type": "application/json",
                "X-Channel-Id": "SDK NodeJS V 1.01",
                Authorization: `Bearer ${this.accessToken}`,
            };
            axios_1.default
                .get(`${url}/${constant_1.ENDPOINT.ECOMMERCE_BUYER_TRANSACTION}?transaction=${numberOfTransaction}`, { headers: headers })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
    requestEcommerceMerchantIdentity() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const url = yield ((_a = this.configuration) === null || _a === void 0 ? void 0 : _a.getAPIUrl());
            const headers = {
                "Content-Type": "application/json",
                "X-Channel-Id": "SDK NodeJS V 1.01",
                Authorization: `Bearer ${this.accessToken}`,
            };
            axios_1.default
                .get(`${url}/${constant_1.ENDPOINT.ECOMMERCE_MERCHANT_IDENTITY}`, { headers: headers })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
    requestEcommerceMerchantIncome(from, to) {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const url = yield ((_a = this.configuration) === null || _a === void 0 ? void 0 : _a.getAPIUrl());
            const headers = {
                "Content-Type": "application/json",
                "X-Channel-Id": "SDK NodeJS V 1.01",
                Authorization: `Bearer ${this.accessToken}`,
            };
            axios_1.default
                .get(`${url}/${constant_1.ENDPOINT.ECOMMERCE_MERCHANT_INCOME}?from=${from}&to=${to}`, { headers: headers })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
    requestEcommerceMerchantTransaction(numberOfTransaction) {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const url = yield ((_a = this.configuration) === null || _a === void 0 ? void 0 : _a.getAPIUrl());
            const headers = {
                "Content-Type": "application/json",
                "X-Channel-Id": "SDK NodeJS V 1.01",
                Authorization: `Bearer ${this.accessToken}`,
            };
            axios_1.default
                .get(`${url}/${constant_1.ENDPOINT.ECOMMERCE_BUYER_MERCHANT_TRANSACTION}?transaction=${numberOfTransaction}`, { headers: headers })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
}
exports.EcommerceDataWorker = EcommerceDataWorker;
