// onebrick-node-sdk/dist/core/worker/eWalletWorker.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EWalletWorker = void 0;
const axios_1 = __importDefault(require("axios"));
const constant_1 = require("../util/constant");
class EWalletWorker {
    constructor(configuration, institution) {
        this.configuration = configuration;
        this.institution = institution;
        this.configuration.setCurrentInstitution(this.institution);
    }
    requestReauthenticationAndLink(pin, urlOvo) {
        var _a, _b, _c, _d, _e, _f;
        let userAuthPayload = {
            deviceId: (_a = this.configuration) === null || _a === void 0 ? void 0 : _a.deviceId,
            refId: (_b = this.configuration) === null || _b === void 0 ? void 0 : _b.refId,
            pin: Number(pin),
            otpNumber: urlOvo,
            redirectRefId: (_d = (_c = this.configuration) === null || _c === void 0 ? void 0 : _c.authRequestData) === null || _d === void 0 ? void 0 : _d.data.redirectRefId,
            username: (_e = this.configuration) === null || _e === void 0 ? void 0 : _e.username,
        };
        let instName = "ovo";
        if (((_f = this.institution) === null || _f === void 0 ? void 0 : _f.toString()) === "11") {
            instName = "gopay";
        }
        console.log(userAuthPayload);
        console.log(instName);
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _g, _h, _j, _k, _l, _m;
            const url = yield ((_g = this === null || this === void 0 ? void 0 : this.configuration) === null || _g === void 0 ? void 0 : _g.getAPIUrl());
            console.log(`${url}/${constant_1.ENDPOINT.AUTH_USER}/${instName}/${(_j = (_h = this.configuration) === null || _h === void 0 ? void 0 : _h.authRequestData) === null || _j === void 0 ? void 0 : _j.data.clientId}?redirectRefId=${userAuthPayload.redirectRefId}&uuid=true`);
            axios_1.default
                .post(`${url}/${constant_1.ENDPOINT.AUTH_USER}/${instName}/${(_l = (_k = this.configuration) === null || _k === void 0 ? void 0 : _k.authRequestData) === null || _l === void 0 ? void 0 : _l.data.clientId}?redirectRefId=${userAuthPayload.redirectRefId}&uuid=true`, JSON.stringify(userAuthPayload), {
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${(_m = this.configuration) === null || _m === void 0 ? void 0 : _m.publicAccessToken}`,
                },
            })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
    requestReAuthentication(otp) {
        var _a, _b, _c, _d, _e, _f, _g;
        let userAuthPayload = {
            institutionId: this.institution,
            username: (_a = this.configuration) === null || _a === void 0 ? void 0 : _a.username,
            otp: otp,
            sessionId: (_b = this.configuration) === null || _b === void 0 ? void 0 : _b.sessionId,
            uniqueId: (_c = this.configuration) === null || _c === void 0 ? void 0 : _c.uniqueId,
            otpToken: (_d = this.configuration) === null || _d === void 0 ? void 0 : _d.otpToken,
            redirectRefId: (_f = (_e = this.configuration) === null || _e === void 0 ? void 0 : _e.authRequestData) === null || _f === void 0 ? void 0 : _f.data.redirectRefId,
        };
        console.log(userAuthPayload);
        let instName = "ovo";
        if (((_g = this.institution) === null || _g === void 0 ? void 0 : _g.toString()) === "11") {
            instName = "gopay";
        }
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _h, _j, _k, _l;
            const url = yield ((_h = this === null || this === void 0 ? void 0 : this.configuration) === null || _h === void 0 ? void 0 : _h.getAPIUrl());
            axios_1.default
                .post(`${url}/${constant_1.ENDPOINT.AUTH_USER}/${instName}/${(_k = (_j = this.configuration) === null || _j === void 0 ? void 0 : _j.authRequestData) === null || _k === void 0 ? void 0 : _k.data.clientId}?redirectRefId=${userAuthPayload.redirectRefId}`, JSON.stringify(userAuthPayload), {
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${(_l = this.configuration) === null || _l === void 0 ? void 0 : _l.publicAccessToken}`,
                },
            })
                .then((data) => {
                resolve(data.data);
            })
                .catch((err) => {
                console.log(err);
                reject(err);
            });
        }));
    }
    authenticateUser(phoneNumber) {
        var _a, _b;
        let userAuthPayload = {
            institutionId: this.institution,
            username: phoneNumber,
            redirectRefId: (_b = (_a = this.configuration) === null || _a === void 0 ? void 0 : _a.authRequestData) === null || _b === void 0 ? void 0 : _b.data.redirectRefId,
        };
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _c, _d, _e, _f;
            const url = yield ((_c = this === null || this === void 0 ? void 0 : this.configuration) === null || _c === void 0 ? void 0 : _c.getAPIUrl());
            axios_1.default
                .post(`${url}/${constant_1.ENDPOINT.AUTH_USER}/${(_e = (_d = this.configuration) === null || _d === void 0 ? void 0 : _d.authRequestData) === null || _e === void 0 ? void 0 : _e.data.clientId}`, JSON.stringify(userAuthPayload), {
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${(_f = this.configuration) === null || _f === void 0 ? void 0 : _f.publicAccessToken}`,
                },
            })
                .then((data) => {
                var _a, _b, _c, _d, _e, _f;
                console.log('data 1-->', data);
                (_a = this.configuration) === null || _a === void 0 ? void 0 : _a.setRefId(data.data.data.refId ? data.data.data.refId : "");
                (_b = this.configuration) === null || _b === void 0 ? void 0 : _b.setDeviceId(data.data.data.deviceId ? data.data.data.deviceId : "");
                (_c = this.configuration) === null || _c === void 0 ? void 0 : _c.setOTPToken(data.data.data.otpToken);
                (_d = this.configuration) === null || _d === void 0 ? void 0 : _d.setUniqueId(data.data.data.uniqueId);
                (_e = this.configuration) === null || _e === void 0 ? void 0 : _e.setSessionId(data.data.data.sessionId);
                (_f = this.configuration) === null || _f === void 0 ? void 0 : _f.setCurrentUserName(data.data.data.username);
                resolve(data.data);
            })
                .catch((err) => {
                reject(err);
            });
        }));
    }
}
exports.EWalletWorker = EWalletWorker;
