// onebrick-node-sdk/dist/core/worker/incomeVerification.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncomeVerificationWorker = void 0;
const axios_1 = __importDefault(require("axios"));
const constant_1 = require("../util/constant");
class IncomeVerificationWorker {
    constructor(configuration, institution) {
        this.configuration = configuration;
        this.institution = institution;
        this.configuration.setCurrentInstitution(this.institution);
    }
    authenticateUser(username, password) {
        var _a;
        let userAuthPayload = {
            institutionId: this.institution,
            username: username,
            password: password,
            redirectRefId: (_a = this.configuration.authRequestData) === null || _a === void 0 ? void 0 : _a.data.redirectRefId,
        };
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _b, _c, _d;
            const url = yield ((_b = this === null || this === void 0 ? void 0 : this.configuration) === null || _b === void 0 ? void 0 : _b.getAPIUrl());
            axios_1.default
                .post(`${url}/${constant_1.ENDPOINT.AUTH_USER}/${(_c = this.configuration.authRequestData) === null || _c === void 0 ? void 0 : _c.data.clientId}`, JSON.stringify(userAuthPayload), {
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${(_d = this.configuration) === null || _d === void 0 ? void 0 : _d.publicAccessToken}`,
                },
            })
                .then((data) => {
                console.log(JSON.stringify(data.data));
                resolve(data.data);
            })
                .catch((err) => {
                console.log(err);
                reject(err);
            });
        }));
    }
}
exports.IncomeVerificationWorker = IncomeVerificationWorker;
