// onebrick-node-sdk/dist/core/util/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = void 0;
const index_1 = require("../../index");
const constant_1 = require("./constant");
class Configuration {
    constructor(clientKey, clientSecret, environment, name, redirectURL) {
        this.clientKey = "";
        this.clientSecret = "";
        this.name = "";
        this.cookies = [];
        this.uniqueId = "";
        this.deviceId = "";
        this.sessionId = "";
        this.redirectRefId = "";
        this.refId = "";
        this.password = "";
        this.requestId = "";
        this.username = "";
        this.otpToken = "";
        this.redirectURL = "";
        this.environment = index_1.ENVIRONMENT.SANBOX;
        this.publicAccessToken = "";
        this.clientKey = clientKey;
        this.clientSecret = clientSecret;
        this.redirectURL = redirectURL ? redirectURL : "";
        this.name = name;
        this.environment = environment ? environment : index_1.ENVIRONMENT.SANBOX;
        this.publicAccessToken = "";
    }
    setUserAuthenticationRequestData(authData) {
        this.authRequestData = authData;
    }
    sertRequestId(requestId) {
        this.requestId = requestId;
    }
    setUniqueId(uniqueId) {
        this.uniqueId = uniqueId;
    }
    setSessionId(sessionId) {
        this.sessionId = sessionId;
    }
    setOTPToken(otpToken) {
        this.otpToken = otpToken;
    }
    setRefId(refId) {
        this.refId = refId;
    }
    setDeviceId(deviceId) {
        this.deviceId = deviceId;
    }
    setpassword(password) {
        this.password = password;
    }
    setCurrentUserName(username) {
        this.username = username;
    }
    setCurrentInstitution(insitution) {
        this.currentInstitution = insitution;
    }
    setPublicAccessToken(accessToken) {
        this.publicAccessToken = accessToken;
    }
    setRedirectRefId(refId) {
        this.redirectRefId = refId;
    }
    getAmplitudeKey() {
        return this.environment == index_1.ENVIRONMENT.PRODUCTION
            ? constant_1.CONFIGURATION.AMP_PROD
            : constant_1.CONFIGURATION.AMP_SANDBOX;
    }
    getAPIUrl() {
        return this.environment == index_1.ENVIRONMENT.PRODUCTION
            ? constant_1.CONFIGURATION.CORE_PRODUCTION_API_URL_V1
            : constant_1.CONFIGURATION.CORE_SANDBOX_API_URL_V1;
    }
}
exports.Configuration = Configuration;
