// onebrick-node-sdk/dist/core/util/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENDPOINT = exports.RESPONSE_ERROR = exports.CONFIGURATION = void 0;
var CONFIGURATION;
(function (CONFIGURATION) {
    CONFIGURATION["CORE_SANDBOX_API_URL_V1"] = "https://sandbox.onebrick.io";
    CONFIGURATION["CORE_PRODUCTION_API_URL_V1"] = "https://api.onebrick.io/";
    CONFIGURATION["AMP_PROD"] = "d0494808544696fca1aa166b1575ff76";
    CONFIGURATION["AMP_SANDBOX"] = "f195edc9416b5214d2d7cfe7da104929";
    CONFIGURATION["SOURCE"] = "sdk-node";
})(CONFIGURATION = exports.CONFIGURATION || (exports.CONFIGURATION = {}));
var RESPONSE_ERROR;
(function (RESPONSE_ERROR) {
    RESPONSE_ERROR["INVALID_PHONE_NUMBER"] = "Invalid Phone number format";
})(RESPONSE_ERROR = exports.RESPONSE_ERROR || (exports.RESPONSE_ERROR = {}));
var ENDPOINT;
(function (ENDPOINT) {
    ENDPOINT["REQUEST_ACCESS_TOKEN"] = "v1/auth/token";
    ENDPOINT["INSTITUTION_LIST"] = "v1/institution";
    ENDPOINT["AUTH_USER"] = "v1/auth";
    ENDPOINT["AUTH_BUSSINESS"] = "v1/business/auth";
    ENDPOINT["ACCOUNT_LIST"] = "v1/account/list";
    ENDPOINT["EMPLOYMENT_DATA_SALARY"] = "v1/income/salary";
    ENDPOINT["EMPLOYMENT_DATA_GENERAL"] = "v1/income/general";
    ENDPOINT["EMPLOYMENT_DATA_EMPLOYMENT"] = "v1/income/employment";
    ENDPOINT["ACCOUNT"] = "v1/account";
    ENDPOINT["ACCOUNT_DATA_GENERAL"] = "v1/account/list";
    ENDPOINT["ACCOUNT_DATA_GENERAL_MFA"] = "v1/account-mfa/list";
    ENDPOINT["TRANSACTION_DATA_GENERAL"] = "v1/transaction/list";
    ENDPOINT["TRANSACTION_DATA_GENERAL_MFA"] = "v1/transaction-mfa/list";
    ENDPOINT["ACCOUNT_DATA_DETAIL_GENERAL"] = "v1/account/list";
    ENDPOINT["ECOMMERCE_BUYER_IDENTITY"] = "v1/ecommerce/identity";
    ENDPOINT["ECOMMERCE_BUYER_TRANSACTION"] = "v1/ecommerce/transaction";
    ENDPOINT["ECOMMERCE_MERCHANT_IDENTITY"] = "v1/ecommerce/merchant-identity";
    ENDPOINT["ECOMMERCE_MERCHANT_INCOME"] = "v1/ecommerce/merchant-income";
    ENDPOINT["ECOMMERCE_BUYER_MERCHANT_TRANSACTION"] = "v1/ecommerce/merchant-transaction";
    ENDPOINT["INSIGHT_AVG_BALANCE"] = "v1/avg-balance";
    ENDPOINT["INSIGHT_BALANCE_SUMMARY"] = "v1/transaction-balance-summary";
    ENDPOINT["INSIGHT_CATEGORIZATION"] = "v1/categorization";
    ENDPOINT["STATEMENT_BASE"] = "v1/statement";
    ENDPOINT["STATEMENT_DOCUMENT_EXTRACT"] = "v1/documents/extracts";
})(ENDPOINT = exports.ENDPOINT || (exports.ENDPOINT = {}));
var TRACKING;
(function (TRACKING) {
})(TRACKING || (TRACKING = {}));
